### ISSUE TITLE
Refactor post-*-data and post-* views/Cannot structure posts

### ISSUE THREAD

**Issue by Angamanga** [ushahidi/platform#2116]
### Expected behaviour
- Post detail view should use a single directive
- Post edit views should use a single directive

### Actual behaviour
Post-detail.controller is replicated, except for sending post as a dependency and setting layout in this file
https://github.com/ushahidi/platform-client/blob/28dc5a488c6c11053c870de4e1406425ee4b84a4/app/main/posts/detail/post-detail-data.directive.js
Same for edit etc

Should be refactored so both can use the same directive

Causing validation problems:
When in post detail view, I select edit and then try to add the post to a survey, but I get the validation error and the http response is `422 Unprocessable Entity`
![edit-post-detail-view-error](https://user-images.githubusercontent.com/8941223/32195281-3d3a8460-bd93-11e7-9895-7fa5e49528b8.gif)

### Steps to reproduce the behaviour/error

**Comment by rjmackay** [comment]
@jshorland @justinscherer Should there even be a standalone post detail views? or should post detail just always load the data sidebar alongside?

**Comment by justinscherer** [comment]
@rjmackay 

> Should there even be a standalone post detail views? or should post detail just always load the data sidebar alongside?

In mobile it opens in a new view because 2-pane doesn't work at that viewport size.

**Comment by rjmackay** [comment]
Hrm ok. That should probably be driven by the same view though, just with responsive CSS hiding the rest of the data view. (or something like that)

Right now if you get to an individual post via the map it'll open a standalone view, even on desktop

**Comment by justinscherer** [comment]
> you get to an individual post via the map it'll open a standalone view, even on desktop

you're right. Sorry, I was justing thinking about the data view...

Having the sidebar beside/on top of the map view would be weird, imo. 

Ideally we could just leave it inconsistent for now and reassess post-election.

**Comment by rjmackay** [comment]
> Having the sidebar beside/on top of the map view would be weird, imo.

I'm not really sure what you mean here. How would we have a sidebar beside the map view?

In case I wasn't clear, if you follow these steps:
1. Open map view
2. Click on a post marker
3. Click on the post title
... You end up on the detail with with neither the map or the data view sidebar.

**Comment by justinscherer** [comment]
I've been chewing on this since yesterday. I don't mind if these go to data view for now. It's not even that weird because the current behaviour is to jump to a separate view. 

To sum up:

1. open map view
2. click on post marker
3. click on post title
...system navigates to data view with this postcard selected. 

The only question in my mind here is what else appears in the postcard list on the left side of the page when this post is highlighted? 

@crcommons @rowasc @rjmackay @Angamanga :

What would be the obvious default behaviour of the postcard list on the left?

Does this fix our single directive issue?

**Comment by crcommons** [comment]
Refactor spec:

- Remove detail data view completely
- Individual post view should default to data view of post

When you either 1) click on a post from the map or 2) go to a post link (e.g. uchaguzi.or.ke/posts/post_ID), it should:
- Route to data view with the post selected at the top of the list and post detail open on the right side
- It should load/show the selected post first and then go get the next 19 posts in the DB (taking into account any filters) in order to ensure the user isn't waiting extra time for all 20 posts to show

**Comment by rjmackay** [comment]
That looks sane to me. The fiddly bit will be ensure we load all the relevant posts in the postcard list, not just the first page. But thats solvable

**Comment by rowasc** [comment]
This is looking good. 
Questions and some assumptions regarding: 
- "It should load/show the selected post first and then go get the next 19 posts in the DB (taking into account any filters) in order to ensure the user isn't waiting extra time for all 20 posts to show"

-----

- `Assumption 1 is that this means:` "show the 19 posts we would have returned given the set of filters we have and the sorting options we have, but with this one (selected) guaranteed to be on top and selected"? 
is that correct? 
- `Assumption 2: `  when they hit next page/load more, the originally selected post remains on top. 
- `Question:` should we somehow highlight the fact that the first post is the one they selected? 
-  `Question about Filters/Apply filters: ` a user selected a post, we're in the data view with that selected post, and the next 19 posts load. Now they change the filters, and the first post doesn't even match so it should not be displayed. Is that the expected behavior, or do you think the user would expect to be able to keep the post there  ("sticky" ?)  even as they make changes?  



@justinscherer @rjmackay @Angamanga @crcommons

**Comment by justinscherer** [comment]
> Assumption 1 is that this means: "show the 19 posts we would have returned given the set of filters we have and the sorting options we have, but with this one (selected) guaranteed to be on top and selected"?
is that correct?

correct

> Assumption 2: when they hit next page/load more, the originally selected post remains on top.

correct

> Question: should we somehow highlight the fact that the first post is the one they selected?

It's nudged in and visible in the right pane, so it uses the same on-state as any postcard. no need for a special on-state. 

> Question about Filters/Apply filters: a user selected a post, we're in the data view with that selected post, and the next 19 posts load. Now they change the filters, and the first post doesn't even match so it should not be displayed. Is that the expected behavior, or do you think the user would expect to be able to keep the post there ("sticky" ?) even as they make changes?

that is the expected behaviour, yes. Postcards are never sticky.

**Comment by rowasc** [comment]
status update:

>It should load/show the selected post first and then go get the next 19 posts in the DB (taking into account any filters) in order to ensure the user isn't waiting extra time for all 20 posts to show 

- If you go to the URL directly (*) we show you the post before anything loads
- After that, the 19 next posts load

> when they hit next page/load more, the originally selected post remains on top.
- works on the new router migration branches 

TODO: 
* the URL for posts in the data view is `/views/data/posts/6547` instead of `/posts/6547` in the router, which I forgot to update while doing the uirouter migration. @jshorland can we get together in the morning to discuss this?  I'd prefer it if we can keep it like this but if not I'll look for a workaround with the routes with a redirect.

**Comment by rowasc** [comment]
@jshorland see comment below

**Comment by Angamanga** [comment]
Status-update: I have fixed the styles for making the data-view responsive so that:
1. when you go to data-view on a small screen, the post-list only is visible
2. when you click a post, this post only becomes visible, and there is a 'x' that should close the post and go back to the post-list again.

When this is working, the stand-alone-detail-view can be removed.

Started a pr for client-changes here: https://github.com/ushahidi/platform-client/pull/930 and @crcommons is continuing the changes.

**Comment by crcommons** [comment]
Currently sitting in uirouter-migration-init branch. Once that is merged into Develop, this is ready for QA.

**Comment by Angamanga** [comment]
Passes QA
